src/blockchain/state_transition_function.cpp

```diff
@@ -0,0 +1,274 @@
+                                const Justifications &justifications) {
+    auto &roots = state.justifications_roots.data();
+    auto &validators = state.justifications_validators.data();
+    roots.resize(0);
+    roots.reserve(justifications.size());
+    validators.resize(0);
+    validators.reserve(justifications.size() * VALIDATOR_REGISTRY_LIMIT);
+    for (auto &[root, bits] : justifications) {
+      BOOST_ASSERT(bits.size() == VALIDATOR_REGISTRY_LIMIT);
+      roots.push_back(root);
+      validators.insert(validators.end(), bits.begin(), bits.end());
+    }
+  }
+
+  State STF::generateGenesisState(const Config &config) const {
+    BlockHeader header;
+    header.body_root = sszHash(BlockBody{});
+    return State{
+        .config = config,
+        .latest_block_header = header,
+    };
+  }
+
+  Block STF::genesisBlock(const State &state) const {
+    return Block{.state_root = sszHash(state)};
+  }
+
+  outcome::result<State> STF::stateTransition(const SignedBlock &signed_block,
+                                              const State &parent_state,
+                                              bool check_state_root) const {
+    auto &block = signed_block.message;
+    auto state = parent_state;
+    // Process slots (including those with no blocks) since block
+    OUTCOME_TRY(processSlots(state, block.slot));
+    // Process block
+    OUTCOME_TRY(processBlock(state, block));
+    // Verify state root
+    if (check_state_root) {
+      auto state_root = sszHash(state);
+      if (block.state_root != state_root) {
+        return Error::STATE_ROOT_DOESNT_MATCH;
+      }
+    }
+    return std::move(state);
+  }
+
+  outcome::result<void> STF::processSlots(State &state, Slot slot) const {
+    if (state.slot >= slot) {
+      return Error::INVALID_SLOT;
+    }
+    while (state.slot < slot) {
+      processSlot(state);
+      ++state.slot;
+    }
+    return outcome::success();
+  }
+
+  void STF::processSlot(State &state) const {
+    // Cache latest block header state root
+    if (state.latest_block_header.state_root == kZeroHash) {
+      state.latest_block_header.state_root = sszHash(state);
+    }
+  }
+
+  outcome::result<void> STF::processBlock(State &state,
+                                          const Block &block) const {
+    OUTCOME_TRY(processBlockHeader(state, block));
+    OUTCOME_TRY(processOperations(state, block.body));
+    return outcome::success();
+  }
+
+  outcome::result<void> STF::processBlockHeader(State &state,
+                                                const Block &block) const {
+    // Verify that the slots match
+    if (block.slot != state.slot) {
+      return Error::INVALID_SLOT;
+    }
+    // Verify that the block is newer than latest block header
+    if (block.slot <= state.latest_block_header.slot) {
+      return Error::INVALID_SLOT;
+    }
+    // Verify that proposer index is the correct index
+    if (block.proposer_index != block.slot % state.config.num_validators) {
```

Comment:
Probably, it should be extract to separated function, i.e. 'validateProposerIndex'